File: services/budget-service/internal/transport/http/metrics.go
package http

import (
	"fmt"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
)

type metricStore struct {
	mu       sync.Mutex
	requests map[string]int64
	latency  map[string]durationSummary
}

type durationSummary struct {
	count int64
	sum   float64
}

var appMetrics = metricStore{
	requests: make(map[string]int64),
	latency:  make(map[string]durationSummary),
}

func metricsMiddleware() gin.HandlerFunc {
	return func(ctx *gin.Context) {
		if ctx.Request.URL.Path == "/metrics" {
			ctx.Next()
			return
		}

		startedAt := time.Now()
		ctx.Next()

		route := ctx.FullPath()
		if route == "" {
			route = ctx.Request.URL.Path
		}
		method := ctx.Request.Method
		status := ctx.Writer.Status()
		elapsed := time.Since(startedAt).Seconds()

		requestKey := fmt.Sprintf("%s\x00%s\x00%d", method, route, status)
		latencyKey := fmt.Sprintf("%s\x00%s", method, route)

		appMetrics.mu.Lock()
		appMetrics.requests[requestKey]++
		current := appMetrics.latency[latencyKey]
		current.count++
		current.sum += elapsed
		appMetrics.latency[latencyKey] = current
		appMetrics.mu.Unlock()
	}
}

func metricsHandler(ctx *gin.Context) {
	appMetrics.mu.Lock()
	defer appMetrics.mu.Unlock()

	var lines []string
	lines = append(lines,
		"# HELP app_http_requests_total Total HTTP requests handled by the service.",
		"# TYPE app_http_requests_total counter",
	)

	requestKeys := make([]string, 0, len(appMetrics.requests))
	for key := range appMetrics.requests {
		requestKeys = append(requestKeys, key)
	}
	sort.Strings(requestKeys)
	for _, key := range requestKeys {
		parts := strings.Split(key, "\x00")
		lines = append(lines, fmt.Sprintf(
			"app_http_requests_total{service=\"budget-service\",method=\"%s\",route=\"%s\",status=\"%s\"} %d",
			parts[0],
			parts[1],
			parts[2],
			appMetrics.requests[key],
		))
	}

	lines = append(lines,
		"# HELP app_http_request_duration_seconds HTTP request latency summary.",
		"# TYPE app_http_request_duration_seconds summary",
	)

	latencyKeys := make([]string, 0, len(appMetrics.latency))
	for key := range appMetrics.latency {
		latencyKeys = append(latencyKeys, key)
	}
	sort.Strings(latencyKeys)
	for _, key := range latencyKeys {
		parts := strings.Split(key, "\x00")
		labels := fmt.Sprintf("service=\"budget-service\",method=\"%s\",route=\"%s\"", parts[0], parts[1])
		summary := appMetrics.latency[key]
		lines = append(lines, fmt.Sprintf("app_http_request_duration_seconds_count{%s} %d", labels, summary.count))
		lines = append(lines, fmt.Sprintf("app_http_request_duration_seconds_sum{%s} %.6f", labels, summary.sum))
	}

	ctx.Data(http.StatusOK, "text/plain; version=0.0.4", []byte(strings.Join(lines, "\n")+"\n"))
}

File: services/auth-service/core/metrics.py
import threading
import time
from collections import defaultdict

from django.http import HttpResponse


REQUESTS = defaultdict(int)
LATENCY_SECONDS = defaultdict(lambda: {"count": 0.0, "sum": 0.0})
LOCK = threading.Lock()
SERVICE_NAME = "auth-service"


class MetricsMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        if request.path == "/metrics":
            return self.get_response(request)

        started_at = time.monotonic()
        response = self.get_response(request)
        elapsed_seconds = time.monotonic() - started_at

        route = getattr(getattr(request, "resolver_match", None), "route", None) or request.path
        key = (request.method, route, str(response.status_code))
        with LOCK:
            REQUESTS[key] += 1
            summary = LATENCY_SECONDS[(request.method, route)]
            summary["count"] += 1
            summary["sum"] += elapsed_seconds

        return response


def metrics_view(_request):
    lines = [
        "# HELP app_http_requests_total Total HTTP requests handled by the service.",
        "# TYPE app_http_requests_total counter",
    ]
    with LOCK:
        for (method, route, status), value in sorted(REQUESTS.items()):
            lines.append(
                'app_http_requests_total{service="%s",method="%s",route="%s",status="%s"} %d'
                % (SERVICE_NAME, method, route, status, value)
            )

        lines.extend(
            [
                "# HELP app_http_request_duration_seconds HTTP request latency summary.",
                "# TYPE app_http_request_duration_seconds summary",
            ]
        )
        for (method, route), summary in sorted(LATENCY_SECONDS.items()):
            count = int(summary["count"])
            total = summary["sum"]
            labels = 'service="%s",method="%s",route="%s"' % (
                SERVICE_NAME,
                method,
                route,
            )
            lines.append(f"app_http_request_duration_seconds_count{{{labels}}} {count}")
            lines.append(f"app_http_request_duration_seconds_sum{{{labels}}} {total:.6f}")

    return HttpResponse("\n".join(lines) + "\n", content_type="text/plain; version=0.0.4")
